runtime/vm/xz64/UpcallThunkGen.cpp

```diff
@@ -0,0 +1,762 @@
+	instructionPtr[4] = ((D2 & 0xFF000) >> 12);
+	instructionPtr[5] = 0x66;
+	return instructionSize;
+}
+
+static I_32
+LAY(I_8 *instructionPtr, I_32 R1, I_32 X2, I_32 B2, I_32 D2)
+{
+	// LAY R1,D2(X2,B2)
+	I_32 instructionSize = 6;
+	*((I_32 *)instructionPtr) = 0xE3000000 | ((R1 & 0xF) << 20) | ((X2 & 0xF) << 16) | ((B2 & 0xF) << 12) | (D2 & 0xFFF);
+	instructionPtr[4] = ((D2 & 0xFF000) >> 12);
+	instructionPtr[5] = 0x71;
+	return instructionSize;
+}
+
+static I_32
+BASR(I_8 *instructionPtr, I_32 R1, I_32 R2)
+{
+	// BASR R1,R2
+	I_32 instructionSize = 2;
+	I_32 opcode = 0x0D00;
+	*((I_16 *)instructionPtr) = (opcode | (((I_16)R1) << 4) | (((I_16)R2)));
+	return instructionSize;
+}
+
+static I_32
+IIHF(I_8 *instructionPtr, I_32 R1, I_32 I2)
+{
+	// IIHF R1, I2
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xC0080000;
+	*((I_32 *)instructionPtr) = (opcode | (R1 << 20));
+	instructionPtr += 2;
+	*((I_32 *)instructionPtr) = I2;
+	return instructionSize;
+}
+
+static I_32
+IILF(I_8 *instructionPtr, I_32 R1, I_32 I2)
+{
+	// IIHF R1, I2 or IILF R1, I2
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xC0090000;
+	*((I_32 *)instructionPtr) = (opcode | (R1 << 20));
+	instructionPtr += 2;
+	*((I_32 *)instructionPtr) = I2;
+	return instructionSize;
+}
+
+static I_32
+AGFI(I_8 *instructionPtr, I_32 R1, I_32 I2)
+{
+	// "AGFI R1,I2"
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xC2080000;
+	*((I_32 *)instructionPtr) = (opcode | (R1 << 20) | ((I2 & 0xFFFF0000) >> 16));
+	instructionPtr[4] = (I_8)((0x0000FF00 & I2) >> 8);
+	instructionPtr[5] = (I_8)(0x000000FF & I2);
+	return instructionSize;
+}
+
+static I_32
+LG(I_8 *instructionPtr, I_32 R1, I_32 X2, I_32 B2, I_32 D2)
+{
+	// "LG R1,D2(X2,B2)"
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xE3000000;
+	*((I_32 *)instructionPtr) = (opcode | (R1 << 20) | (X2 << 16) | (B2 << 12) | (D2 & 0x00000FFF));
+	instructionPtr[4] = (I_8)(D2 & 0x000FFF00);
+	instructionPtr[5] = 0x04;
+	return instructionSize;
+}
+
+static I_32
+STG(I_8 *instructionPtr, I_32 R1, I_32 X2, I_32 B2, I_32 D2)
+{
+	// STG "R1, D2(X2,B2)"
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xE3000000;
+	*((I_32 *)instructionPtr) = (opcode | (R1 << 20) | (X2 << 16) | (B2 << 12) | (D2 & 0x00000FFF));
+	instructionPtr[4] = (I_8)((D2 & 0x000FF000) >> 12);
+	instructionPtr[5] = 0x24;
+	return instructionSize;
+}
+
+static I_32
+MVC(I_8 *instructionPtr, I_32 D1, I_32 L, I_32 B1, I_32 D2, I_32 B2)
+{
+	// MVC D1(L,B1),D2(B2)
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xD2000000;
+	*((I_32 *)instructionPtr) = (opcode | ((L & 0xFF) << 16) | ((B1 & 0xF) << 12) | (D1 & 0xFFF));
+	instructionPtr[4] = ((0x0F & B2) << 4) | ((D2 & 0xF00) >> 8);
+	instructionPtr[5] = D2 & 0xFF;
+	return instructionSize;
+}
+
+static I_32
+generateStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea)
+{
+	I_32 totalInstructionSize = 0;
+	I_32 maxBytes = 255;
```

Comment:
So in revisiting this area I have a couple of observations:

- I don't think the issue mentioned above created any functional issues because the displacement integer calculation accounted for this discrepancy. This probably explains why I didn't see any failures in my test case because I do have edge cases for this scenario. Anyway, you're right that we should set maxBytes to 256. I'll have this fixed in the next commit.

- There is however a bug if we try to copy a structure larger than 4096 bytes. This is because the displacement integer in the instruction is only 12 bits in length. So in scenarios where we have to copy more than 4096 bytes of data, I propose the following solution: 
  - Copy the structure using consecutive MVC instruction for the first 4096 bytes
  - Copy the remaining bytes using an MVC instruction in a loop

The above solution is a small extension of the current method, and I imagine we won't have many practical scenarios where this could cause a performance concern. 

An alternative is to consider MVCL instructions but there are limitations there as well with respect to the displacement integer and for now I don't think the setup work required for that is worth it.

Reply:
Good catch about the 4096 bytes, 
Your proposed solution means we will use 16 MVCs and for the remaining bytes use `MCVL/MVCLE` . Means we will use 100 bytes of instructions with that. Apart from that, you have to emit instruction to increase both source and destination by 4096 and also load the remaining bytes in the register for MVCLE.
Another caveat with using `MVCLE` looking at the PoPs, you would have to also insert branch instruction to account for interruption in copying from source to destination to account for operation completion because of CPU determined number of bytes have been moved without reaching the end of the first operand in which case condition code would be set. 
We should gauge if this is beneficial or simple loop of MVC (MVC, update src, update dst, check residue is less than 256 and branch back).

Reply:
Also I need to confirm this from the instruction documents but the comment in [1] suggests a very high threshold should be used for deciding MVCL. 

[1]. https://github.com/eclipse/omr/blob/e26f652a53fef350a52dd999f2b1ce9a422ab2f0/compiler/z/codegen/OMRTreeEvaluator.cpp#L12813-L12818

Reply:
This routine is now split into two. One copies small structs using upto 16 MVC instructions, and the other one copies larger structs in a loop with 16 MVC instructions at a time. I believe I'm missing one more corner case here. I'll update this comment once I've addressed that.